File: shared/frontend/utils/dataClient.js
class DataClient {
  constructor(gatewayUrl = "http://127.0.0.1:8000") {
    this.base = gatewayUrl.replace(/\/$/, "");
    this.namespace = "__unknown__";
  }
  async create(resource, data) {
    const r = await fetch(`${this.base}/_api/data/${this.namespace}/${resource}`, {
      method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(data)
    });
    if (!r.ok) throw new Error(`create failed: ${r.status}`);
    return r.json();
  }
  async list(resource, { limit = 100, offset = 0 } = {}) {
    const r = await fetch(`${this.base}/_api/data/${this.namespace}/${resource}?limit=${limit}&offset=${offset}`);
    if (!r.ok) throw new Error(`list failed: ${r.status}`);
    return r.json();
  }
  async get(resource, id) {
    const r = await fetch(`${this.base}/_api/data/${this.namespace}/${resource}/${id}`);
    if (!r.ok) throw new Error(`get failed: ${r.status}`);
    return r.json();
  }
  async update(resource, id, data) {
    const r = await fetch(`${this.base}/_api/data/${this.namespace}/${resource}/${id}`, {
      method: "PUT", headers: { "Content-Type": "application/json" }, body: JSON.stringify(data)
    });
    if (!r.ok) throw new Error(`update failed: ${r.status}`);
    return r.json();
  }
  async delete(resource, id) {
    const r = await fetch(`${this.base}/_api/data/${this.namespace}/${resource}/${id}`, {
      method: "DELETE"
    });
    if (!r.ok) throw new Error(`delete failed: ${r.status}`);
    return r.json();
  }
}

File: shared/backend/data_client.py
import os
import httpx


class DataClient:
    def __init__(self, gateway_url: str | None = None):
        self.base = (gateway_url or os.environ.get("GATEWAY_URL", "http://127.0.0.1:8000")).rstrip("/")
        self.namespace = os.environ.get("TOOL_NAME", "unknown")

    def _url(self, resource: str, record_id: int | None = None) -> str:
        u = f"{self.base}/_api/data/{self.namespace}/{resource}"
        return f"{u}/{record_id}" if record_id is not None else u

    def create(self, resource: str, data: dict) -> dict:
        resp = httpx.post(self._url(resource), json=data, timeout=10)
        resp.raise_for_status()
        return resp.json()

    def list(self, resource: str, limit=100, offset=0) -> list[dict]:
        resp = httpx.get(self._url(resource), params={"limit": limit, "offset": offset}, timeout=10)
        resp.raise_for_status()
        return resp.json()

    def get(self, resource: str, record_id: int) -> dict:
        resp = httpx.get(self._url(resource, record_id), timeout=10)
        resp.raise_for_status()
        return resp.json()

    def update(self, resource: str, record_id: int, data: dict) -> dict:
        resp = httpx.put(self._url(resource, record_id), json=data, timeout=10)
        resp.raise_for_status()
        return resp.json()

    def delete(self, resource: str, record_id: int) -> dict:
        resp = httpx.delete(self._url(resource, record_id), timeout=10)
        resp.raise_for_status()
        return resp.json()
